File: .skills/openclaw-skills/skills/0xnyk/xint/lib/stream.ts
/**
 * stream.ts — Official X filtered stream + stream rules management.
 *
 * Uses:
 *   GET  /2/tweets/search/stream
 *   GET  /2/tweets/search/stream/rules
 *   POST /2/tweets/search/stream/rules
 */

import * as api from "./api";
import * as fmt from "./format";
import { checkBudget, trackCost } from "./costs";
import { buildOutputMeta } from "./output-meta";
import { validateWebhookUrl } from "./webhook-security";

interface StreamOptions {
  json?: boolean;
  jsonl?: boolean;
  maxEvents?: number;
  backfillMinutes?: number;
  webhook?: string;
  quiet?: boolean;
}

interface StreamRule {
  id: string;
  value: string;
  tag?: string;
}

const RULES_URL = `${api.BASE}/tweets/search/stream/rules`;

export function parsePositiveInt(value: string | undefined): number | null {
  if (!value) return null;
  const n = Number.parseInt(value, 10);
  if (!Number.isFinite(n) || n <= 0) return null;
  return n;
}

export function parseBackfillMinutes(value: string | undefined): number | null {
  const n = parsePositiveInt(value);
  if (n === null) return null;
  if (n < 1 || n > 5) return null;
  return n;
}

function nowIso(): string {
  return new Date().toISOString().replace("T", " ").slice(0, 19);
}

async function postWebhook(url: string, payload: any): Promise<void> {
  try {
    await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
  } catch (err: any) {
    console.error(`[webhook] Failed: ${err?.message || String(err)}`);
  }
}

async function listRules(): Promise<StreamRule[]> {
  const raw = await api.bearerGet(RULES_URL);
  const rules = Array.isArray(raw?.data) ? raw.data : [];
  return rules as StreamRule[];
}

async function addRule(value: string, tag?: string): Promise<any> {
  const body: any = { add: [{ value }] };
  if (tag) body.add[0].tag = tag;
  return api.bearerPost(RULES_URL, body);
}

async function deleteRules(ids: string[]): Promise<any> {
  return api.bearerPost(RULES_URL, { delete: { ids } });
}

function printStreamHelp(): void {
  console.log(`
Usage: xint stream [options]

Open official X filtered stream and print matching tweets in real time.
Manage stream rules with: xint stream-rules ...

Options:
  --json                 Output structured JSON per event
  --jsonl                Output compact JSONL per event
  --max-events N         Stop after N tweet events
  --backfill N           Backfill 1-5 minutes (X API option)
  --webhook <url>        POST each event payload to webhook (https:// required)
  --quiet, -q            Suppress stream status logs

Webhook security:
  - Remote webhooks must use https://
  - http:// is allowed only for localhost/127.0.0.1/::1
  - Optional host allowlist: XINT_WEBHOOK_ALLOWED_HOSTS=hooks.example.com,*.internal.example

Examples:
  xint stream
  xint stream --jsonl
  xint stream --backfill 2 --max-events 25
  xint stream --webhook https://hooks.example.com/ingest
`);
}

function printStreamRulesHelp(): void {
  console.log(`
Usage: xint stream-rules <subcommand> [options]

Subcommands:
  list [--json]                          List current filtered-stream rules (default)
  add <rule expression> [--tag <tag>]    Add a rule
  delete <rule_id...>                    Delete one or more rule IDs
  clear                                  Delete all rules

Examples:
  xint stream-rules
  xint stream-rules add "from:elonmusk -is:retweet" --tag elon
  xint stream-rules delete 1770000012345678901
  xint stream-rules clear
`);
}

async function runStream(opts: StreamOptions): Promise<void> {
  const token = api.getBearerToken();
  const backfill = opts.backfillMinutes ? `&backfill_minutes=${opts.backfillMinutes}` : "";
  const url = `${api.BASE}/tweets/search/stream?${api.FIELDS}${backfill}`;

  const controller = new AbortController();
  let running = true;
  let eventsSeen = 0;
  const startMs = Date.now();

  const onSignal = () => {
    running = false;
    controller.abort();
  };
  process.on("SIGINT", onSignal);
  process.on("SIGTERM", onSignal);

  try {
    if (!opts.quiet) {
      console.error("Connecting to X filtered stream...");
      console.error("Press Ctrl+C to stop");
    }

    const res = await fetch(url, {
      headers: { Authorization: `Bearer ${token}` },
      signal: controller.signal,
    });

    if (res.status === 429) {
      const reset = res.headers.get("x-rate-limit-reset");
      const waitSec = reset
        ? Math.max(Number.parseInt(reset, 10) - Math.floor(Date.now() / 1000), 1)
        : 60;
      throw new Error(`Rate limited. Resets in ${waitSec}s`);
    }

    if (!res.ok) {
      const text = await res.text();
      throw new Error(`X API ${res.status}: ${text.slice(0, 200)}`);
    }

    if (!res.body) throw new Error("No response body from stream endpoint.");

    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    while (running) {
      const { done, value } = await reader.read();
      if (done) break;
      if (!value) continue;

      buffer += decoder.decode(value, { stream: true });
      let newlineIndex = buffer.indexOf("\n");
      while (newlineIndex >= 0) {
        const rawLine = buffer.slice(0, newlineIndex);
        buffer = buffer.slice(newlineIndex + 1);
        newlineIndex = buffer.indexOf("\n");

        const line = rawLine.trim();
        if (!line) continue; // keep-alive newline

        let payload: any;
        try {
          payload = JSON.parse(line);
        } catch {
          continue;
        }
        if (!payload?.data) continue;

        const tweets = api.parseTweets(payload as any);
        if (tweets.length === 0) continue;

        trackCost("stream_connect", "/2/tweets/search/stream", tweets.length);

        for (const tweet of tweets) {
          eventsSeen += 1;
          const event = {
            timestamp: new Date().toISOString(),
            matching_rules: payload.matching_rules || [],
            tweet,
          };

          const meta = buildOutputMeta({
            source: "x_api_v2",
            startedAtMs: startMs,
            cached: false,
            confidence: 1,
            apiEndpoint: "/2/tweets/search/stream",
            estimatedCostUsd: 0.005,
          });

          if (opts.json) {
            console.log(JSON.stringify({ meta, data: event }, null, 2));
          } else if (opts.jsonl) {
            console.log(JSON.stringify({ ...meta, event }));
          } else {
            if (!opts.quiet) {
              console.error(`[${nowIso()}] Stream match (${event.matching_rules.length} rule(s))`);
            }
            console.log(fmt.formatTweetTelegram(tweet, undefined, { full: false }));
            console.log();
          }

          if (opts.webhook) {
            await postWebhook(opts.webhook, { source: "xint-stream", ...event });
          }

          const budget = checkBudget();
          if (!budget.allowed) {
            console.error(
              `\n!! Budget exceeded ($${budget.spent.toFixed(2)}/$${budget.limit.toFixed(2)}). Stopping stream.`,
            );
            running = false;
            break;
          }

          if (opts.maxEvents && eventsSeen >= opts.maxEvents) {
            running = false;
            break;
          }
        }
      }
    }
  } catch (err: any) {
    // Ignore abort errors from Ctrl+C.
    if (!(err?.name === "AbortError")) {
      throw err;
    }
  } finally {
    process.off("SIGINT", onSignal);
    process.off("SIGTERM", onSignal);
    const elapsed = Math.round((Date.now() - startMs) / 1000);
    const mins = Math.floor(elapsed / 60);
    const secs = elapsed % 60;
    if (!opts.quiet) {
      console.error(`\n--- Stream stopped ---`);
      console.error(`Duration: ${mins}m ${secs}s | Events: ${eventsSeen}`);
    }
  }
}

export async function cmdStream(args: string[]): Promise<void> {
  let json = false;
  let jsonl = false;
  let quiet = false;
  let maxEvents: number | undefined;
  let backfillMinutes: number | undefined;
  let webhook: string | undefined;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    switch (arg) {
      case "--json":
        json = true;
        break;
      case "--jsonl":
        jsonl = true;
        break;
      case "--quiet":
      case "-q":
        quiet = true;
        break;
      case "--max-events": {
        const parsed = parsePositiveInt(args[++i]);
        if (parsed === null) {
          throw new Error("Usage: --max-events requires a positive integer.");
        }
        maxEvents = parsed;
        break;
      }
      case "--backfill": {
        const parsed = parseBackfillMinutes(args[++i]);
        if (parsed === null) {
          throw new Error("Usage: --backfill requires an integer from 1 to 5.");
        }
        backfillMinutes = parsed;
        break;
      }
      case "--webhook":
        webhook = args[++i];
        if (!webhook) throw new Error("Usage: --webhook requires a URL.");
        break;
      case "--help":
      case "-h":
        printStreamHelp();
        return;
      default:
        throw new Error(`Unknown option for stream: ${arg}`);
    }
  }

  if (json && jsonl) {
    throw new Error("Use only one of --json or --jsonl.");
  }

  if (webhook) {
    webhook = validateWebhookUrl(webhook);
  }

  await runStream({ json, jsonl, quiet, maxEvents, backfillMinutes, webhook });
}

export async function cmdStreamRules(args: string[]): Promise<void> {
  let json = false;
  let tag: string | undefined;
  const parts: string[] = [];

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    switch (arg) {
      case "--json":
        json = true;
        break;
      case "--tag":
        tag = args[++i];
        if (!tag) throw new Error("Usage: --tag requires a value.");
        break;
      case "--help":
      case "-h":
        printStreamRulesHelp();
        return;
      default:
        parts.push(arg);
        break;
    }
  }

  const sub = (parts[0] || "list").toLowerCase();

  if (sub === "list") {
    const rules = await listRules();
    trackCost("stream_rules_list", "/2/tweets/search/stream/rules", 0);

    if (json) {
      console.log(JSON.stringify(rules, null, 2));
      return;
    }

    if (rules.length === 0) {
      console.log("No stream rules configured.");
      return;
    }

    console.log(`\n📡 Stream Rules (${rules.length})\n`);
    for (let i = 0; i < rules.length; i++) {
      const rule = rules[i];
      const tagText = rule.tag ? ` [${rule.tag}]` : "";
      console.log(`${i + 1}. ${rule.id}${tagText}`);
      console.log(`   ${rule.value}`);
    }
    return;
  }

  if (sub === "add") {
    const value = parts.slice(1).join(" ").trim();
    if (!value) {
      throw new Error("Usage: xint stream-rules add <rule expression> [--tag <tag>]");
    }
    const res = await addRule(value, tag);
    trackCost("stream_rules_add", "/2/tweets/search/stream/rules", 0);

    if (json) {
      console.log(JSON.stringify(res, null, 2));
      return;
    }
    const added = Array.isArray(res?.data) ? res.data.length : 0;
    console.log(`✅ Added ${added || 1} stream rule(s).`);
    return;
  }

  if (sub === "delete") {
    const ids = parts.slice(1);
    if (ids.length === 0) {
      throw new Error("Usage: xint stream-rules delete <rule_id...>");
    }
    const res = await deleteRules(ids);
    trackCost("stream_rules_delete", "/2/tweets/search/stream/rules", 0);

    if (json) {
      console.log(JSON.stringify(res, null, 2));
      return;
    }
    console.log(`✅ Deleted ${ids.length} stream rule(s).`);
    return;
  }

  if (sub === "clear") {
    const rules = await listRules();
    if (rules.length === 0) {
      console.log("No stream rules to clear.");
      return;
    }
    const ids = rules.map((r) => r.id);
    const res = await deleteRules(ids);
    trackCost("stream_rules_delete", "/2/tweets/search/stream/rules", 0);

    if (json) {
      console.log(JSON.stringify(res, null, 2));
      return;
    }
    console.log(`✅ Cleared ${ids.length} stream rule(s).`);
    return;
  }

  throw new Error(`Unknown stream-rules subcommand: ${sub}`);
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/commands/stream.rs
use anyhow::{bail, Result};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

use crate::api::twitter;
use crate::cli::{StreamArgs, StreamRulesArgs};
use crate::client::{XClient, FIELDS};
use crate::config::Config;
use crate::costs;
use crate::format;
use crate::models::RawResponse;
use crate::output_meta;
use crate::webhook::validate_webhook_url;

#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
struct StreamRule {
    id: String,
    value: String,
    tag: Option<String>,
}

fn now_display() -> String {
    chrono::Utc::now().format("%Y-%m-%d %H:%M:%S").to_string()
}

fn validate_backfill(minutes: Option<u8>) -> Result<Option<u8>> {
    match minutes {
        Some(v) if (1..=5).contains(&v) => Ok(Some(v)),
        Some(_) => bail!("--backfill must be between 1 and 5."),
        None => Ok(None),
    }
}

async fn fetch_rules(client: &XClient, token: &str) -> Result<Vec<StreamRule>> {
    let raw = client
        .bearer_get("tweets/search/stream/rules", token)
        .await?;
    let data = match raw.data {
        Some(serde_json::Value::Array(arr)) => serde_json::Value::Array(arr),
        Some(_) | None => serde_json::Value::Array(Vec::new()),
    };
    Ok(serde_json::from_value(data).unwrap_or_default())
}

pub async fn run_stream(args: &StreamArgs, config: &Config, client: &XClient) -> Result<()> {
    if args.json && args.jsonl {
        bail!("Use only one of --json or --jsonl.");
    }

    let token = config.require_bearer_token()?;
    let backfill = validate_backfill(args.backfill)?;
    let webhook_url = match args.webhook.as_deref() {
        Some(raw) => Some(validate_webhook_url(raw)?),
        None => None,
    };

    let mut path = format!("tweets/search/stream?{FIELDS}");
    if let Some(minutes) = backfill {
        path.push_str(&format!("&backfill_minutes={minutes}"));
    }

    let running = Arc::new(AtomicBool::new(true));
    let r = running.clone();
    ctrlc::set_handler(move || {
        r.store(false, Ordering::SeqCst);
    })?;

    if !args.quiet {
        eprintln!("Connecting to X filtered stream...");
        eprintln!("Press Ctrl+C to stop");
    }

    let mut response = client.bearer_stream(&path, token).await?;
    let mut buffer = String::new();
    let start_time = std::time::Instant::now();
    let mut events_seen: usize = 0;

    while running.load(Ordering::SeqCst) {
        let chunk = response.chunk().await?;
        let Some(bytes) = chunk else { break };
        if bytes.is_empty() {
            continue;
        }

        buffer.push_str(&String::from_utf8_lossy(&bytes));
        while let Some(newline_idx) = buffer.find('\n') {
            let line = buffer[..newline_idx].trim().to_string();
            buffer = buffer[newline_idx + 1..].to_string();
            if line.is_empty() {
                continue;
            }

            let value: serde_json::Value = match serde_json::from_str(&line) {
                Ok(v) => v,
                Err(_) => continue,
            };
            if value.get("data").is_none() {
                continue;
            }

            let raw: RawResponse = match serde_json::from_value(value.clone()) {
                Ok(v) => v,
                Err(_) => continue,
            };
            let tweets = twitter::parse_tweets(&raw);
            if tweets.is_empty() {
                continue;
            }

            costs::track_cost(
                &config.costs_path(),
                "stream_connect",
                "/2/tweets/search/stream",
                tweets.len() as u64,
            );

            for tweet in tweets {
                events_seen += 1;
                let matching_rules = value
                    .get("matching_rules")
                    .cloned()
                    .unwrap_or_else(|| serde_json::json!([]));
                let tweet_json = serde_json::to_value(&tweet)?;
                let event = serde_json::json!({
                    "timestamp": chrono::Utc::now().to_rfc3339(),
                    "matching_rules": matching_rules,
                    "tweet": tweet_json,
                });

                let meta = output_meta::build_meta(
                    "x_api_v2",
                    start_time,
                    false,
                    1.0,
                    "/2/tweets/search/stream",
                    0.005,
                    &config.costs_path(),
                );

                if args.json {
                    output_meta::print_json_with_meta(&meta, &event)?;
                } else if args.jsonl {
                    let payload = serde_json::json!({
                        "source": meta.source,
                        "latency_ms": meta.latency_ms,
                        "cached": meta.cached,
                        "confidence": meta.confidence,
                        "api_endpoint": meta.api_endpoint,
                        "timestamp": meta.timestamp,
                        "estimated_cost_usd": meta.estimated_cost_usd,
                        "budget_remaining_usd": meta.budget_remaining_usd,
                        "event": event
                    });
                    println!("{}", serde_json::to_string(&payload)?);
                } else {
                    if !args.quiet {
                        let rule_count = event["matching_rules"]
                            .as_array()
                            .map(|a| a.len())
                            .unwrap_or(0);
                        eprintln!("[{}] Stream match ({} rule(s))", now_display(), rule_count);
                    }
                    println!("{}", format::format_tweet_terminal(&tweet, None, false));
                    println!();
                }

                if let Some(webhook) = &webhook_url {
                    if let Err(err) = client
                        .post_json(
                            webhook,
                            &serde_json::json!({
                                "source": "xint-stream",
                                "event": event,
                            }),
                        )
                        .await
                    {
                        eprintln!("[webhook] Failed: {err}");
                    }
                }

                let budget = costs::check_budget(&config.costs_path());
                if !budget.allowed {
                    eprintln!(
                        "\n!! Budget exceeded (${:.2}/${:.2}). Stopping stream.",
                        budget.spent, budget.limit
                    );
                    running.store(false, Ordering::SeqCst);
                    break;
                }

                if let Some(max_events) = args.max_events {
                    if events_seen >= max_events {
                        running.store(false, Ordering::SeqCst);
                        break;
                    }
                }
            }
        }
    }

    let elapsed = start_time.elapsed().as_secs();
    if !args.quiet {
        eprintln!("\n--- Stream stopped ---");
        eprintln!(
            "Duration: {}m {}s | Events: {}",
            elapsed / 60,
            elapsed % 60,
            events_seen
        );
    }

    Ok(())
}

pub async fn run_stream_rules(
    args: &StreamRulesArgs,
    config: &Config,
    client: &XClient,
) -> Result<()> {
    let token = config.require_bearer_token()?;
    let parts = args
        .subcommand
        .clone()
        .unwrap_or_else(|| vec!["list".to_string()]);
    let sub = parts[0].to_lowercase();

    match sub.as_str() {
        "list" => {
            let rules = fetch_rules(client, token).await?;
            costs::track_cost(
                &config.costs_path(),
                "stream_rules_list",
                "/2/tweets/search/stream/rules",
                0,
            );

            if args.json {
                format::print_json_pretty_filtered(&rules)?;
                return Ok(());
            }

            if rules.is_empty() {
                println!("No stream rules configured.");
                return Ok(());
            }

            println!("\nStream Rules ({})\n", rules.len());
            for (idx, rule) in rules.iter().enumerate() {
                let tag = rule.tag.clone().unwrap_or_default();
                let tag_display = if tag.is_empty() {
                    String::new()
                } else {
                    format!(" [{tag}]")
                };
                println!("{}. {}{}", idx + 1, rule.id, tag_display);
                println!("   {}", rule.value);
            }
        }
        "add" => {
            let value = parts[1..].join(" ").trim().to_string();
            if value.is_empty() {
                bail!("Usage: xint stream-rules add <rule expression> [--tag <tag>]");
            }

            let mut rule_obj = serde_json::json!({ "value": value });
            if let Some(tag) = &args.tag {
                rule_obj["tag"] = serde_json::Value::String(tag.clone());
            }

            let body = serde_json::json!({ "add": [rule_obj] });
            let res = client
                .bearer_post("tweets/search/stream/rules", token, Some(&body))
                .await?;
            costs::track_cost(
                &config.costs_path(),
                "stream_rules_add",
                "/2/tweets/search/stream/rules",
                0,
            );

            if args.json {
                format::print_json_pretty_filtered(&res)?;
            } else {
                println!("✅ Added stream rule.");
            }
        }
        "delete" => {
            let ids: Vec<String> = parts[1..].iter().map(ToString::to_string).collect();
            if ids.is_empty() {
                bail!("Usage: xint stream-rules delete <rule_id...>");
            }

            let body = serde_json::json!({ "delete": { "ids": ids } });
            let res = client
                .bearer_post("tweets/search/stream/rules", token, Some(&body))
                .await?;
            costs::track_cost(
                &config.costs_path(),
                "stream_rules_delete",
                "/2/tweets/search/stream/rules",
                0,
            );

            if args.json {
                format::print_json_pretty_filtered(&res)?;
            } else {
                println!("✅ Deleted stream rule(s).");
            }
        }
        "clear" => {
            let rules = fetch_rules(client, token).await?;
            if rules.is_empty() {
                println!("No stream rules to clear.");
                return Ok(());
            }
            let ids: Vec<String> = rules.into_iter().map(|r| r.id).collect();
            let body = serde_json::json!({ "delete": { "ids": ids } });
            let res = client
                .bearer_post("tweets/search/stream/rules", token, Some(&body))
                .await?;
            costs::track_cost(
                &config.costs_path(),
                "stream_rules_delete",
                "/2/tweets/search/stream/rules",
                0,
            );
            if args.json {
                format::print_json_pretty_filtered(&res)?;
            } else {
                println!("✅ Cleared all stream rules.");
            }
        }
        _ => {
            bail!("Unknown stream-rules subcommand: {sub}");
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::validate_backfill;

    #[test]
    fn validate_backfill_accepts_valid_range() {
        assert_eq!(
            validate_backfill(Some(1)).expect("expected valid lower bound"),
            Some(1)
        );
        assert_eq!(
            validate_backfill(Some(5)).expect("expected valid upper bound"),
            Some(5)
        );
    }

    #[test]
    fn validate_backfill_rejects_invalid_values() {
        assert!(validate_backfill(Some(0)).is_err());
        assert!(validate_backfill(Some(6)).is_err());
        assert_eq!(validate_backfill(None).expect("expected None"), None);
    }
}
